### ISSUE TITLE
Checkpoint with independent data set

### ISSUE THREAD

**Issue by robkuehl** [keras-team/keras#12112]
For a project in University I have been given two data sets.
The first one is the training data set with 200.000 data points, and the second one is the testing set, with 256 data points, which should not be used in fitting.
Now, I want to save the network, which is performing best on the test data during the fitting.
So I guess I want to implement a callback working with the test data set, which is independent from the the validation split.
How can I do that?

**Comment by Dref360** [comment]
The validation set should come from the same distribution as the test set. If there is a big performance gap between them, it means that they do not come from the same distribition.

You should **never** use the test set beside for reporting results. Otherwise, you would just use the test set as the validation set and use EarlyStopping. If you do this, you would indirectly optimize on the test set and this is not right.

**Comment by AdministradorViejo** [comment]
I'm not sure if it is clear what I want to do.
I just want to measure the performance of the current network after each epoch on a data set, which is not the validation set used for the loss function, and save weights for the best performance.
I asked my lecturer, he said I am allowed to that.
The problem is, I don't know how to implement that, as loading a validation_set into the fit function just overwrites the validation_split.

**Comment by longw010** [comment]
In my opinion, you could write a customized Callback class, and override the `on_epoch_end()` function (where you could run `predict` etc).

**Comment by AdministradorViejo** [comment]
Ok, should it work, if I overwrite the ModelCheckpoint Class and just change the monitor?
I loaded the data into the class and calculate the monitor with self.model.evaluate.

**Comment by tRosenflanz** [comment]
As @Dref360 said this will be a leakage. Your lecturer might have misunderstood you. 

Proper practice is to fit on training data and save model that performs the best on validation. Repeat this multiple times until you get an acceptable performance on validation set and only then run the saved model against the test set. Iterating on the test set is a big "no no". So in your case you should not check against 256 test points until you are done fitting the model. Ideally you only touch test case once/twice during the whole project.

But yes, you can easily do that using a custom callback that checks calculates the metric you want on the some holdout data and saves the model if it is better than the previous best
